File: src/ts/reactive.ts
import {
  BehaviorSubject,
  Subject,
  distinctUntilChanged,
  filter,
  map,
  type Observable,
} from "rxjs";

import {
  type DndEnvelope,
  type DndLifecyclePhase,
  type DndOperation,
  type DndTelemetryEvent,
  telemetryFor,
  validateEnvelope,
} from "./index.js";

/**
 * Process-local reactive event. The envelope may contain dragged data, so this
 * type must not be sent to telemetry, persisted as replay history, or exposed
 * through analytics/crash-reporting adapters.
 */
export interface DndReactiveEvent {
  readonly phase: DndLifecyclePhase;
  readonly envelope: DndEnvelope;
  readonly operation: DndOperation | null;
  readonly targetId: string | null;
}

/**
 * Replay-safe reactive state. Deliberately contains metadata only: no item
 * payload, media contents, form values, or other dragged `data` field.
 */
export interface DndReactiveState {
  readonly active: boolean;
  readonly phase: DndLifecyclePhase | null;
  readonly dragId: string | null;
  readonly sourceRuntime: string | null;
  readonly itemCount: number;
  readonly operation: DndOperation | null;
  readonly targetId: string | null;
}

export const IDLE_DND_REACTIVE_STATE: DndReactiveState = Object.freeze({
  active: false,
  phase: null,
  dragId: null,
  sourceRuntime: null,
  itemCount: 0,
  operation: null,
  targetId: null,
});

const START_PHASES: ReadonlySet<DndLifecyclePhase> = new Set(["drag-start", "drag-enter"]);
const LOSSLESS_PHASES: ReadonlySet<DndLifecyclePhase> = new Set(["drop", "drag-end"]);

/** `drag-over` may be sampled/coalesced for presentation work. */
export function isHighFrequencyLifecyclePhase(phase: DndLifecyclePhase): boolean {
  return phase === "drag-over";
}

/** `drop` and `drag-end` are terminal-significant and must never be throttled away. */
export function isLosslessLifecyclePhase(phase: DndLifecyclePhase): boolean {
  return LOSSLESS_PHASES.has(phase);
}

export function reactiveStateFor(event: DndReactiveEvent): DndReactiveState {
  return Object.freeze({
    active: event.phase !== "drag-end",
    phase: event.phase,
    dragId: event.envelope.dragId,
    sourceRuntime: event.envelope.sourceRuntime,
    itemCount: event.envelope.items.length,
    operation: event.operation,
    targetId: event.targetId,
  });
}

function assertReactiveEventSemantics(event: DndReactiveEvent): void {
  if (event.operation !== null && !event.envelope.allowedOperations.includes(event.operation)) {
    throw new Error("reactive event operation is not source-allowed");
  }
  if (event.targetId !== null && event.targetId.length === 0) {
    throw new Error("reactive event targetId must be a non-empty string");
  }
  if (event.phase === "drop") {
    if (event.operation === null) {
      throw new Error("drop event requires a negotiated operation");
    }
    if (event.targetId === null) {
      throw new Error("drop event requires a targetId");
    }
  }
}

function canTransition(
  previous: DndLifecyclePhase | null,
  next: DndLifecyclePhase,
): boolean {
  if (previous === null) return START_PHASES.has(next);
  switch (previous) {
    case "drag-start":
      return next === "drag-enter" || next === "drag-over" || next === "drop" || next === "drag-end";
    case "drag-enter":
      return next === "drag-over" || next === "drag-leave" || next === "drop" || next === "drag-end";
    case "drag-over":
      return next === "drag-over" || next === "drag-leave" || next === "drop" || next === "drag-end";
    case "drag-leave":
      return next === "drag-enter" || next === "drag-over" || next === "drag-end";
    case "drop":
      return next === "drag-end";
    case "drag-end":
      return START_PHASES.has(next);
  }
}

/**
 * Fail-closed lifecycle tracker shared by browser/webview consumers.
 *
 * A session may begin with `drag-start` for an in-app drag or `drag-enter` for
 * an external/system drag. Once active, events may not silently switch drag IDs.
 * `drop` must be followed only by `drag-end` before a new session begins.
 */
export class DndLifecycleTracker {
  #dragId: string | null = null;
  #phase: DndLifecyclePhase | null = null;

  get dragId(): string | null {
    return this.#dragId;
  }

  get phase(): DndLifecyclePhase | null {
    return this.#phase;
  }

  accept(event: DndReactiveEvent): DndReactiveState {
    assertReactiveEventSemantics(event);

    if (this.#dragId !== null && event.envelope.dragId !== this.#dragId) {
      throw new Error(`reactive dragId switched before drag-end: ${this.#dragId} -> ${event.envelope.dragId}`);
    }
    if (!canTransition(this.#phase, event.phase)) {
      throw new Error(`invalid reactive lifecycle transition: ${this.#phase ?? "idle"} -> ${event.phase}`);
    }

    const state = reactiveStateFor(event);
    if (event.phase === "drag-end") {
      this.#dragId = null;
      this.#phase = null;
    } else {
      this.#dragId = event.envelope.dragId;
      this.#phase = event.phase;
    }
    return state;
  }

  reset(): void {
    this.#dragId = null;
    this.#phase = null;
  }
}

export interface EmitDndReactiveOptions {
  readonly operation?: DndOperation;
  readonly targetId?: string;
}

/**
 * RxJS-backed hot event bus for drag/drop lifecycles.
 *
 * Raw events use a plain Subject and therefore do not replay dragged payloads.
 * Only the metadata-only state stream is replayed with a BehaviorSubject.
 * Persistence/forms remain explicit `commitAcceptedDrop` effects and are never
 * triggered by `emit`.
 */
export class OresDndReactiveBus {
  readonly #events = new Subject<DndReactiveEvent>();
  readonly #state = new BehaviorSubject<DndReactiveState>(IDLE_DND_REACTIVE_STATE);
  readonly #tracker = new DndLifecycleTracker();

  readonly events$: Observable<DndReactiveEvent> = this.#events.asObservable();
  readonly state$: Observable<DndReactiveState> = this.#state.asObservable();
  readonly active$: Observable<boolean> = this.state$.pipe(
    map((state) => state.active),
    distinctUntilChanged(),
  );
  readonly drops$: Observable<DndReactiveEvent> = this.events$.pipe(
    filter((event) => event.phase === "drop"),
  );
  readonly dragOvers$: Observable<DndReactiveEvent> = this.events$.pipe(
    filter((event) => isHighFrequencyLifecyclePhase(event.phase)),
  );
  readonly lossless$: Observable<DndReactiveEvent> = this.events$.pipe(
    filter((event) => isLosslessLifecyclePhase(event.phase)),
  );
  readonly telemetry$: Observable<DndTelemetryEvent> = this.events$.pipe(
    map((event) => telemetryFor(
      event.phase,
      event.envelope,
      event.operation ?? undefined,
      event.targetId ?? undefined,
    )),
  );

  emit(
    phase: DndLifecyclePhase,
    envelope: DndEnvelope,
    options: EmitDndReactiveOptions = {},
  ): void {
    const safeEnvelope = validateEnvelope(envelope);
    const event: DndReactiveEvent = Object.freeze({
      phase,
      envelope: safeEnvelope,
      operation: options.operation ?? null,
      targetId: options.targetId ?? null,
    });
    const state = this.#tracker.accept(event);
    this.#events.next(event);
    this.#state.next(state);
  }

  complete(): void {
    this.#events.complete();
    this.#state.complete();
    this.#tracker.reset();
  }
}

File: src/rust/src/reactive.rs
use std::convert::Infallible;

use crate::{
    telemetry_for, DndEnvelope, DndError, DndLifecyclePhase, DndOperation, DndTelemetryEvent,
    ValidationOptions,
};
use rxrust::prelude::Local;
use rxrust::ObservableFactory;

/// Re-export rxRust's operator/context prelude from the ORES reactive surface.
/// UI/WASM callers normally choose `Local`; cross-thread native callers may
/// opt into `Shared` explicitly.
pub use rxrust::prelude as rx;

/// Process-local reactive event. The envelope can contain dragged data, so this
/// event must never be serialized into telemetry, replay logs, analytics, or
/// crash-reporting state.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DndReactiveEvent {
    pub phase: DndLifecyclePhase,
    pub envelope: DndEnvelope,
    pub operation: Option<DndOperation>,
    pub target_id: Option<String>,
}

impl DndReactiveEvent {
    pub fn new(
        phase: DndLifecyclePhase,
        envelope: DndEnvelope,
        operation: Option<DndOperation>,
        target_id: Option<String>,
    ) -> Result<Self, DndError> {
        envelope.validate(ValidationOptions::default())?;
        if let Some(operation) = operation {
            if !envelope.allowed_operations.contains(&operation) {
                return Err(DndError(
                    "reactive event operation is not source-allowed".to_owned(),
                ));
            }
        }
        if target_id.as_deref().is_some_and(str::is_empty) {
            return Err(DndError(
                "reactive event targetId must be a non-empty string".to_owned(),
            ));
        }
        Ok(Self {
            phase,
            envelope,
            operation,
            target_id,
        })
    }
}

/// Replay-safe current state. This deliberately has no DndItem/data field.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DndReactiveState {
    pub active: bool,
    pub phase: Option<DndLifecyclePhase>,
    pub drag_id: Option<String>,
    pub source_runtime: Option<String>,
    pub item_count: usize,
    pub operation: Option<DndOperation>,
    pub target_id: Option<String>,
}

impl DndReactiveState {
    pub fn idle() -> Self {
        Self {
            active: false,
            phase: None,
            drag_id: None,
            source_runtime: None,
            item_count: 0,
            operation: None,
            target_id: None,
        }
    }
}

pub fn reactive_state_for(event: &DndReactiveEvent) -> DndReactiveState {
    DndReactiveState {
        active: event.phase != DndLifecyclePhase::DragEnd,
        phase: Some(event.phase),
        drag_id: Some(event.envelope.drag_id.clone()),
        source_runtime: Some(event.envelope.source_runtime.clone()),
        item_count: event.envelope.items.len(),
        operation: event.operation,
        target_id: event.target_id.clone(),
    }
}

pub fn reactive_telemetry_for(event: &DndReactiveEvent) -> DndTelemetryEvent {
    telemetry_for(
        event.phase,
        &event.envelope,
        event.operation,
        event.target_id.clone(),
    )
}

/// Canonical hot rxRust subject for UI/WASM drag events. It intentionally does
/// not replay envelopes. Derive replayable state by mapping events through
/// [reactive_state_for] into a `Local::behavior_subject` when a UI needs current
/// state semantics.
pub type DndLocalEventSubject =
    rxrust::subject::LocalSubject<'static, DndReactiveEvent, Infallible>;

pub fn local_event_subject() -> DndLocalEventSubject {
    Local::subject::<DndReactiveEvent, Infallible>()
}

#[cfg(test)]
mod tests {
    use std::{cell::RefCell, rc::Rc};

    use super::rx::*;
    use super::*;
    use crate::{DndItem, DndItemKind, ORES_DND_PROTOCOL};

    fn envelope() -> DndEnvelope {
        DndEnvelope {
            protocol: ORES_DND_PROTOCOL.to_owned(),
            drag_id: "drag-rx-1".to_owned(),
            source_runtime: "rust-test".to_owned(),
            allowed_operations: vec![DndOperation::Copy, DndOperation::Move],
            items: vec![DndItem {
                kind: DndItemKind::Text,
                media_type: "text/plain".to_owned(),
                data: "TOP-SECRET-DRAG-DATA".to_owned(),
                name: None,
            }],
            traceparent: None,
            form_id: None,
        }
    }

    #[test]
    fn rxrust_local_pipeline_projects_payload_free_state_and_telemetry() -> Result<(), DndError> {
        let events = vec![
            DndReactiveEvent::new(DndLifecyclePhase::DragStart, envelope(), None, None)?,
            DndReactiveEvent::new(
                DndLifecyclePhase::Drop,
                envelope(),
                Some(DndOperation::Copy),
                Some("zone-a".to_owned()),
            )?,
            DndReactiveEvent::new(
                DndLifecyclePhase::DragEnd,
                envelope(),
                Some(DndOperation::Copy),
                Some("zone-a".to_owned()),
            )?,
        ];

        let states = Rc::new(RefCell::new(Vec::new()));
        let state_sink = Rc::clone(&states);
        Local::from_iter(events.clone())
            .map(|event| reactive_state_for(&event))
            .subscribe(move |state| state_sink.borrow_mut().push(state));

        let telemetry = Rc::new(RefCell::new(Vec::new()));
        let telemetry_sink = Rc::clone(&telemetry);
        Local::from_iter(events)
            .map(|event| reactive_telemetry_for(&event))
            .subscribe(move |event| telemetry_sink.borrow_mut().push(event));

        let states_json = serde_json::to_string(&*states.borrow())?;
        let telemetry_json = serde_json::to_string(&*telemetry.borrow())?;
        assert!(!states_json.contains("TOP-SECRET-DRAG-DATA"));
        assert!(!telemetry_json.contains("TOP-SECRET-DRAG-DATA"));
        assert_eq!(states.borrow().last().map(|state| state.active), Some(false));
        assert_eq!(telemetry.borrow().len(), 3);
        Ok(())
    }

    #[test]
    fn rxrust_subject_is_hot_and_filters_drop_events() -> Result<(), DndError> {
        let mut subject = local_event_subject();
        subject.next(DndReactiveEvent::new(
            DndLifecyclePhase::DragStart,
            envelope(),
            None,
            None,
        )?);

        let drops = Rc::new(RefCell::new(Vec::new()));
        let drop_sink = Rc::clone(&drops);
        subject
            .clone()
            .filter(|event| event.phase == DndLifecyclePhase::Drop)
            .subscribe(move |event| drop_sink.borrow_mut().push(event.phase));

        assert!(drops.borrow().is_empty());
        subject.next(DndReactiveEvent::new(
            DndLifecyclePhase::DragOver,
            envelope(),
            None,
            None,
        )?);
        subject.next(DndReactiveEvent::new(
            DndLifecyclePhase::Drop,
            envelope(),
            Some(DndOperation::Copy),
            Some("zone-a".to_owned()),
        )?);
        assert_eq!(&*drops.borrow(), &[DndLifecyclePhase::Drop]);
        Ok(())
    }
}
